### DISCUSSION TITLE
Is there a way to implement attach-or-create semantics?

### DISCUSSION THREAD

**Discussion by catlee** [zellij-org/zellij#2461] [Q&A]
I'd like to be able to have a hotkey that launches a terminal that attaches to any unattached zellij sessions, or creates a new one if there are no unattached sessions.

`zellij list-sessions` doesn't indicate if a session is attached to or not.

`zellij attach -c` will always attach to a session if it exists.

Is there a way to achieve this?

**Comment by imsnif** [comment]
I'm sorry, I think I don't understand... what's wrong with `zellij attach --create`?

**Reply by catlee** [reply]
What I'd like to happen:
1. Launch first terminal which runs `zellij attach-or-create`. No current sessions exist, so it creates a new one.
2. Launch second terminal which runs `zellij attach-or-create`. The first session is already attached to a terminal, so zellij creates a new one. Now I have two sessions, each in its own terminal.

What happens now:
1. First terminal launches which runs `zellij attach -c`. No current sessions exist, so it creates a new one. So far so good.
2. Second terminal launches which runs `zellij attach -c`. It attaches to the first session. Now I have two terminals attached to the same session.

When I was using tmux, I had a [little helper script](https://github.com/catlee/dotfiles/blob/master/linux/tmux_attach) that accomplished this for me. But AFAICT zellij doesn't expose the right information that would allow me to script this in the same way.

If `zellij list-sessions` included the attached state of the session, then I could write a simple wrapper to select the first unattached session, or to create a new one if no unattached sessions exist.

**Reply by Lockszmith-GH** [reply]
Just run `zellij` without `attach`, it will create a new session.

Attach means - connect to an already existing session.

**Reply by catlee** [reply]
That will always create a new session. I'd like to attach to any detached sessions first, if there are any.

**Reply by Lockszmith-GH** [reply]
ah, I see now - seems I understand fast when explained slow - sorry about that.  
Now I'm interested :)

Although, to be fair, I feel that running multiple - unrelated sessions - on the same machine at random, feels like over management - so don't think I'll use it.
I would rather use panels for that.

If I do need a new session, then I use named sessions, this way, if there is a specific purpose, I name the session with that purpose - so the session is either created or attached. Taking out the guesswork.

My default session name is `$USERNAME@($hostname)` unless I specify one.

Hope you find a solution.

**Reply by catlee** [reply]
Here's a draft PR that implements this idea: https://github.com/zellij-org/zellij/pull/2492

**Reply by Zaubentrucker** [reply]
I use the following bash script as a workaround:

```bash
#!/usr/bin/bash
ZELLIJ_PS_COUNT=$(pgrep zellij | wc -l)

if [ $ZELLIJ_PS_COUNT -eq 1 ]; then
	# `list-sessions` should only give us one item now
	zellij attach $(zellij list-sessions)
else
	zellij
fi
```

This works as long as there are no other zellij sessions running alongside the detached one.
Otherwise this will open a new session even if a detached one is still running.
